plugins/zns/zns.c

```diff
@@ -652,17 +693,37 @@ static int report_zones(int argc, char **argv, struct command *cmd, struct plugi
 		goto close_fd;
 	}
 
-	err = nvme_zns_report_zones(fd, cfg.namespace_id, cfg.zslba,
-		cfg.extended, cfg.state, cfg.partial, report_size, report);
-	if (!err)
-		nvme_show_zns_report_zones(report, cfg.num_descs, zdes,
-			report_size, flags);
-	else if (err > 0)
-		nvme_show_status(err);
-	else
-		perror("zns report-zones");
+	offset = cfg.zslba;
+	printf("nr_zones: %"PRIu64"\n", (uint64_t)le64_to_cpu(total_nr_zones));
+
+    while (nr_zones_retrieved < nr_zones) {
+		if (nr_zones_retrieved >= nr_zones)
+			break;
+
+		if (nr_zones_retrieved + nr_zones_chunks > nr_zones) {
+			nr_zones_chunks = nr_zones - nr_zones_retrieved;
+			log_len = sizeof(struct nvme_zone_report) + ((sizeof(struct nvme_zns_desc) * nr_zones_chunks) + (nr_zones_chunks * zdes));
+		}
+
+		err = nvme_zns_report_zones(fd, cfg.namespace_id, offset,
+			cfg.extended, cfg.state, cfg.partial, log_len, report);
+		if (err > 0) {
+			nvme_show_status(err);
+			break;
+		}
+
+		if (!err)
+			 nvme_show_zns_report_zones(report, nr_zones_chunks, zdes,
```

Comment:
This is going to change the output because it will print out the "nr_zones" header multiple times. Maybe that's okay, though, I'll see how it looks.

Reply:
I moved the print of "nr_zones from the nvme_show_zns_report_zones function into this function so that shouldn't be a problem.